packages/@react-stately/utils/src/useControlledState.ts

```diff
@@ -59,16 +62,19 @@ export function useControlledState<T>(
       setStateValue(updateFunction);
     } else {
       if (!isControlled) {
-        setStateValue(value);
+        setStateValue(newValue);
       }
-      onChangeCaller(value, ...args);
+      onChangeCaller(newValue, ...args);
     }
   }, [isControlled, onChange]);
 
-  // If a controlled component's value prop changes, we need to update stateRef
-  if (isControlled) {
-    stateRef.current = value;
-  } else {
+  useEffect(() => {
+    // If a controlled component's value prop changes, we need to update stateRef
+    if (isControlled) {
+      stateRef.current = value;
+    }
+  });
```

Comment:
I still think this is too late. Imagine we have a controlled component. We fire onChange, this triggers a re-render of the parent component. Then another onChange fires _before_ the useEffect is called, but we would be comparing against the previous value in `stateRef.current`. This is definitely a problem with useEffect because it runs after an animation frame, but I think it's even a problem with useLayoutEffect.

For example, say you have two components: a parent and a child. The parent component has a `useControlledState` in it, and passes its `onChange` function to the child. The child has a useLayoutEffect in it which triggers an onChange in response to something. This is fairly common, for example combobox reacts to prop changes in an effect and might emit events based on that. Same with auto selecting the first item in tabs on mount. The problem is that child effects run before parent effects. If the child component calls onChange in its effect, the value in the ref within the parent component won't have updated yet, meaning the comparison we do would not be correct. You can see an example of that [here](https://codesandbox.io/s/vigilant-kate-q53mdd?file=/src/App.js) - the stateRef is always behind by one render.

Honestly, I'm not really sure how to solve this. For React 18, we _might_ be able to use [useInsertionEffect](https://beta.reactjs.org/reference/react/useInsertionEffect) which runs before layout effects. But you'd still have the same problem if any other component used useInsertionEffect (rare), and for older Reacts.

Updating during render is actually the most close to correct. It's only incorrect when a render gets aborted (e.g. during suspense). For strict mode, the value coming from props should be the same between the two renders so it shouldn't matter. Did you see test failures due to this, or was this just following the linter?

Reply:
Went back and checked the tests passing before and after. It's the same. So changing useControlledState was just following the linter.